// webpack_entry/App.js
import React, { Component } from 'react'
import Lolliplot from '@oncojs/react-lolliplot'
import * as d3 from 'd3'
import mockData from './data'

class App extends Component {
  state = {
    width: 0,
    data: mockData(100),
    min: 0,
    max: 500,
  }
  componentDidMount() {
    this.setState({ width: this.el.clientWidth })
    window.onresize = () => {
      this.setState({ width: this.el.clientWidth })
    }
  }
  update = state => this.setState(state)
  randomize = () => this.setState({ data: mockData(100) })
  reset = () => this.setState({ min: 0, max: 500 })

  render() {
    return (
      <div ref={el => this.el = el}>
        <h1>Lolliplot</h1>
        <input
          type="text"
          value={this.state.min}
          onChange={e => this.setState({ min: +e.target.value })}
        />
        <input
          type="text"
          value={this.state.max}
          onChange={e => this.setState({ max: +e.target.value })}
        />
        {this.state.width &&
          <Lolliplot
            d3={d3}
            update={this.update}
            {...this.state}
          />
        }
        <button onClick={this.randomize}>Randomize</button>
        <button onClick={this.reset}>Reset</button>
      </div>
    )
  }
}

export default App
