// webpack_entry/data.js
export default mutations => ({
  proteins: [
    {
      id: `TAD`,
      start: 0,
      end: 150,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
      getProteinColor: () => `blue`,
    },
    {
      id: `DNA-binding`,
      start: 125,
      end: 350,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
    {
      id: `Oligomerization`,
      start: 325,
      end: 375,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
    {
      id: `NLS`,
      start: 450,
      end: 500,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
  ],
  mutations: [
    ...Array(mutations)
      .fill(1)
      .map((x, i) => ({
        id: `MU${i}`,
        y: Math.round(i * Math.random()),
        x: i * Date.now() % 500,
        consequence: `type_${i % 4 + 1}**`,
        impact:
          Math.random() * 10 < 3.33 ? `HIGH` : Math.random() * 10 > 6.66 ? `MODERATE` : `unknown`,
      }))
      .filter(x => x.y > 0),
  ],
})
