// @oncojs/lolliplot/dist/lib/stats.js
'use strict';

exports.__esModule = true;
exports.updateStats = exports.setupStats = undefined;

var _lodash = require('lodash.startcase');

var _lodash2 = _interopRequireDefault(_lodash);

var _lodash3 = require('lodash.countby');

var _lodash4 = _interopRequireDefault(_lodash3);

var _mutations = require('./mutations');

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

var _animator = require('./animator');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// TODO: place in theme?
var impactsColors = {
  HIGH: 'rgb(221, 60, 60)',
  MODERATE: 'rgb(132, 168, 56)',
  default: 'rgb(135, 145, 150)'
};

var renderCheckboxItem = function renderCheckboxItem(_ref) {
  var colors = _ref.colors,
      type = _ref.type,
      dataSource = _ref.dataSource;
  return '\n  <span\n    class="toggle-checkbox"\n    data-checked="true"\n    style="color: ' + (colors[type] || colors.default) + ';\n    text-align: center;\n    border: 2px solid ' + (colors[type] || colors.default) + ';\n    display: inline-block;\n    width: 23px;\n    cursor: pointer;\n    margin-right: 6px;"\n  >\n    \u2713\n  </span>\n  <span>' + (0, _lodash2.default)(type) + ':</span>\n  <span style="margin: 0 10px 0 auto;" class="counts">\n    <b>' + dataSource[type].length + '</b> / <b>' + dataSource[type].length + '</b>\n  </span>\n';
};

var setupStats = function setupStats(_ref2) {
  var d3 = _ref2.d3,
      d3Root = _ref2.d3Root,
      consequenceColors = _ref2.consequenceColors,
      data = _ref2.data,
      store = _ref2.store,
      hideStats = _ref2.hideStats,
      statsBoxWidth = _ref2.statsBoxWidth,
      width = _ref2.width,
      selectedMutationClass = _ref2.selectedMutationClass,
      consequences = _ref2.consequences,
      impacts = _ref2.impacts,
      mutationChartLines = _ref2.mutationChartLines,
      mutationChartCircles = _ref2.mutationChartCircles,
      selectedMutationBox = _ref2.selectedMutationBox,
      height = _ref2.height,
      xAxisOffset = _ref2.xAxisOffset,
      yTicks = _ref2.yTicks,
      yTicksLine = _ref2.yTicksLine;

  // Stats Bar

  var padding = 35;

  var stats = d3Root.append('div').attr('id', 'mutation-stats').style('display', hideStats ? 'none' : 'block').style('background-color', 'white').style('border', '1px solid rgb(186, 186, 186)').style('padding', '13px').style('width', statsBoxWidth - padding + 'px');

  var mutationCount = stats.style('position', 'absolute').style('top', '0px').style('left', width - statsBoxWidth + padding + 'px').style('line-height', '20px').append('div').html('Viewing <b>' + data.mutations.length + '</b> / <b>' + data.mutations.length + '</b> Mutations').attr('class', 'mutation-count').style('font-size', '16px');

  stats.append('label').attr('for', 'filter-type').text('select filter type').style('display', 'none');

  stats.append('select').attr('id', 'filter-type').html('\n      <option value="Consequence">Consequence</option>\n      <option value="Impact">Impact (VEP)</option>\n    ').style('margin-top', '6px').on('change', function () {
    var isConsequence = d3.event.target.value === 'Consequence';

    var _store$getState = store.getState(),
        consequenceFilters = _store$getState.consequenceFilters,
        impactFilters = _store$getState.impactFilters;

    store.update({ type: d3.event.target.value });
    var filters = isConsequence ? consequenceFilters : impactFilters;

    var mutationClass = isConsequence ? 'consequence' : 'impact';
    var checked = false;
    updateStats({
      d3: d3,
      d3Root: d3Root,
      store: store,
      data: data,
      consequences: consequences,
      impacts: impacts,
      consequenceColors: consequenceColors,
      mutationChartLines: mutationChartLines,
      mutationChartCircles: mutationChartCircles,
      selectedMutationBox: selectedMutationBox,
      height: height,
      xAxisOffset: xAxisOffset,
      consequencesCheckboxContainers: consequencesCheckboxContainers,
      impactsCheckboxContainers: impactsCheckboxContainers,
      mutationCount: mutationCount,
      yTicks: yTicks,
      yTicksLine: yTicksLine,
      type: d3.event.target.value
    });
    (0, _mutations.updateMutations)({ d3Root: d3Root, checked: true, data: data });
    filters.forEach(function (type) {
      return (0, _mutations.updateMutations)({ d3Root: d3Root, checked: checked, mutationClass: mutationClass, type: type, data: data });
    });

    d3Root.selectAll('[id^=class]').style('display', 'none');
    d3Root.select('#class-' + d3.event.target.value).style('display', 'block');

    d3Root.selectAll('[class^=mutation-circle]').attr('fill', function (d) {
      return isConsequence ? consequenceColors[d.consequence] : impactsColors[d.impact] || impactsColors.default;
    });
  });

  stats.append('div').text('Click to filter mutations').style('margin-top', '6px').style('font-size', '11px').style('color', _theme2.default.black);

  var consequencesContainer = stats.append('span').text('Consequence:').attr('id', 'class-Consequence').style('display', selectedMutationClass === 'Consequence' ? 'block' : 'none').style('margin-top', '6px').style('font-size', '14px');

  var consequencesCheckboxContainers = Object.keys(consequences).reduce(function (acc, type) {
    var _Object$assign;

    return Object.assign({}, acc, (_Object$assign = {}, _Object$assign[type] = consequencesContainer.append('div').html(renderCheckboxItem({
      colors: consequenceColors,
      type: type,
      dataSource: consequences
    })).style('margin-top', '6px').style('font-size', '14px').style('display', 'flex').style('align-items', 'center').on('click', function () {
      d3.event.target.dataset.checked = d3.event.target.dataset.checked === 'true' ? 'false' : 'true';

      var checked = d3.event.target.dataset.checked === 'true';

      var _store$getState2 = store.getState(),
          consequenceFilters = _store$getState2.consequenceFilters;

      d3.select(this).select('.toggle-checkbox').html(checked ? '\u2713' : '&nbsp;');

      store.update({ consequenceFilters: checked ? consequenceFilters.filter(function (d) {
          return d !== type;
        }) : [].concat(consequenceFilters, [type])
      });

      updateStats({
        d3: d3,
        d3Root: d3Root,
        store: store,
        data: data,
        consequences: consequences,
        impacts: impacts,
        consequenceColors: consequenceColors,
        mutationChartLines: mutationChartLines,
        mutationChartCircles: mutationChartCircles,
        selectedMutationBox: selectedMutationBox,
        height: height,
        xAxisOffset: xAxisOffset,
        consequencesCheckboxContainers: consequencesCheckboxContainers,
        impactsCheckboxContainers: impactsCheckboxContainers,
        mutationCount: mutationCount,
        yTicks: yTicks,
        yTicksLine: yTicksLine
      });

      (0, _mutations.updateMutations)({ d3Root: d3Root, checked: checked, mutationClass: 'consequence', type: type, data: data });
    }), _Object$assign));
  }, {});

  var impactsContainer = stats.append('span').text('Impact (VEP):').attr('id', 'class-Impact').style('display', selectedMutationClass === 'Impact' ? 'block' : 'none').style('margin-top', '6px').style('font-size', '14px');

  var impactsCheckboxContainers = Object.keys(impacts).reduce(function (acc, type) {
    var _Object$assign2;

    return Object.assign({}, acc, (_Object$assign2 = {}, _Object$assign2[type] = impactsContainer.append('div').html(renderCheckboxItem({
      colors: impactsColors,
      type: type,
      dataSource: impacts
    })).style('margin-top', '6px').style('font-size', '14px').style('display', 'flex').style('align-items', 'center').on('click', function () {
      d3.event.target.dataset.checked = d3.event.target.dataset.checked === 'true' ? 'false' : 'true';

      var checked = d3.event.target.dataset.checked === 'true';

      var _store$getState3 = store.getState(),
          impactFilters = _store$getState3.impactFilters;

      d3.select(this).select('.toggle-checkbox').html(checked ? '\u2713' : '&nbsp;');

      store.update({ impactFilters: checked ? impactFilters.filter(function (d) {
          return d !== type;
        }) : [].concat(impactFilters, [type])
      });

      updateStats({
        d3: d3,
        d3Root: d3Root,
        store: store,
        data: data,
        consequences: consequences,
        impacts: impacts,
        consequenceColors: consequenceColors,
        mutationChartLines: mutationChartLines,
        mutationChartCircles: mutationChartCircles,
        selectedMutationBox: selectedMutationBox,
        height: height,
        xAxisOffset: xAxisOffset,
        consequencesCheckboxContainers: consequencesCheckboxContainers,
        impactsCheckboxContainers: impactsCheckboxContainers,
        mutationCount: mutationCount,
        yTicks: yTicks,
        yTicksLine: yTicksLine
      });

      (0, _mutations.updateMutations)({ d3Root: d3Root, checked: checked, mutationClass: 'impact', type: type, data: data });
    }), _Object$assign2));
  }, {});

  return {
    stats: stats,
    consequencesCheckboxContainers: consequencesCheckboxContainers,
    impactsCheckboxContainers: impactsCheckboxContainers,
    mutationCount: mutationCount
  };
};

var updateStats = function updateStats(_ref3) {
  var d3 = _ref3.d3,
      d3Root = _ref3.d3Root,
      store = _ref3.store,
      data = _ref3.data,
      consequences = _ref3.consequences,
      impacts = _ref3.impacts,
      mutationChartLines = _ref3.mutationChartLines,
      mutationChartCircles = _ref3.mutationChartCircles,
      selectedMutationBox = _ref3.selectedMutationBox,
      height = _ref3.height,
      xAxisOffset = _ref3.xAxisOffset,
      consequencesCheckboxContainers = _ref3.consequencesCheckboxContainers,
      impactsCheckboxContainers = _ref3.impactsCheckboxContainers,
      mutationCount = _ref3.mutationCount,
      yTicks = _ref3.yTicks,
      yTicksLine = _ref3.yTicksLine;

  var _store$getState4 = store.getState(),
      min = _store$getState4.min,
      max = _store$getState4.max,
      consequenceFilters = _store$getState4.consequenceFilters,
      impactFilters = _store$getState4.impactFilters,
      animating = _store$getState4.animating,
      type = _store$getState4.type;

  var visibleMutations = data.mutations.filter(function (d) {
    return d.x > min && d.x < max && (type === 'Consequence' && !consequenceFilters.includes(d.consequence) || type === 'Impact' && !impactFilters.includes(d.impact));
  });

  var visibleMutationCounts = {
    impacts: (0, _lodash4.default)(visibleMutations, 'impact'),
    consequences: (0, _lodash4.default)(visibleMutations, 'consequence')
  };

  Object.entries(consequencesCheckboxContainers).forEach(function (_ref4) {
    var type = _ref4[0],
        container = _ref4[1];
    return container.select('.counts').html('\n        <b>' + (visibleMutationCounts.consequences[type] || 0) + '</b> / <b>' + consequences[type].length + '</b>');
  });

  Object.entries(impactsCheckboxContainers).forEach(function (_ref5) {
    var type = _ref5[0],
        container = _ref5[1];
    return container.select('.counts').html('<b>' + (visibleMutationCounts.impacts[type] || 0) + '</b> / <b>' + impacts[type].length + '</b>');
  });

  mutationCount.html('Viewing <b>' + visibleMutations.length + '</b> / <b>' + data.mutations.length + '</b> Mutations');

  if (!animating) {
    (0, _animator.animateScaleY)({
      d3: d3,
      d3Root: d3Root,
      data: data,
      consequenceFilters: consequenceFilters,
      impactFilters: impactFilters,
      min: min,
      max: max,
      mutationChartLines: mutationChartLines,
      mutationChartCircles: mutationChartCircles,
      selectedMutationBox: selectedMutationBox,
      height: height,
      xAxisOffset: xAxisOffset,
      visibleMutations: visibleMutations,
      yTicks: yTicks,
      yTicksLine: yTicksLine
    });
  }
};

/*----------------------------------------------------------------------------*/

exports.setupStats = setupStats;
exports.updateStats = updateStats;
