// @oncojs/lolliplot/dist/lib/store.js
'use strict';

exports.__esModule = true;

var _redux = require('redux');

var _reduxActions = require('redux-actions');

var createStoreWrapper = function createStoreWrapper(_ref) {
  var domainWidth = _ref.domainWidth;

  var initialState = {
    animating: false,
    min: 0,
    max: domainWidth,
    startMin: 0,
    startMax: domainWidth,
    targetMin: 0,
    targetMax: domainWidth,
    domain: domainWidth,
    currentAnimationIteration: 0,
    consequenceFilters: [],
    impactFilters: [],
    dragging: false,
    zoomStart: null,
    sliding: false,
    slideStart: null,
    slideStartMin: null,
    slideStartMax: null,
    type: 'Consequence'
  };

  var reducer = (0, _reduxActions.handleActions)({
    UPDATE: function UPDATE(state, action) {
      return Object.assign({}, state, action.payload);
    }
  }, initialState);

  var store = (0, _redux.createStore)(reducer);
  store.update = function (payload) {
    return store.dispatch({ type: 'UPDATE', payload: payload });
  };

  return store;
};

exports.default = createStoreWrapper;
