// @oncojs/react-lolliplot/index.js
import React, { Component } from 'react'
import LolliplotNode from './LolliplotNode'

class Lolliplot extends Component {
  state = {
    animating: false,
    startMin: 0,
    startMax: 500,
    targetMin: 0,
    targetMax: 500,
    domain: 500,
    currentAnimationIteration: 0,
    consequenceFilters: [],
    impactFilters: [],
    dragging: false,
    draggingMinimap: false,
    zoomStart: null,
    sliding: false,
    slideStart: null,
    slideStartMin: null,
    slideStartMax: null,
    type: `Consequence`,
    offsetX: 0,
    proteinMouseover: false,
  }

  update = state => this.setState(state)

  render() {
    const lolliplotNodeInstance = (
      <LolliplotNode
        d3={this.props.d3}
        height={450}
        onProteinClick={d => this.props.update({ min: d.start, max: d.end })}
        _update={this.update}
        update={this.props.update}
        onProteinMouseover={d => {
          if (this.props.onProteinMouseover) {
            this.props.onProteinMouseover(d)
          }
          this.update({ proteinMouseover: d.id })
        }}
        onProteinMouseout={d => {
          if (this.props.onProteinMouseout) {
            this.props.onProteinMouseout(d)
          }
          this.update({ proteinMouseover: false })
        }}
        {...this.state}
        {...this.props}
      />
    )

    console.log(lolliplotNodeInstance)
    return lolliplotNodeInstance
  }
}

export default Lolliplot
