// @oncojs/lolliplot/data.js
export default mutations => ({
  proteins: [
    {
      id: `TAD`,
      start: 0,
      end: 75,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
    {
      id: `DNA-binding`,
      start: 125,
      end: 275,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
    {
      id: `Oligomerization`,
      start: 325,
      end: 375,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
    {
      id: `NLS`,
      start: 450,
      end: 500,
      description: `von Hippel-Lindau disease tumour suppressor, beta/alpha domain`,
    },
  ],
  mutations: [
    ...(Array(mutations).fill(1).map((x, i) => ({
      id: `MU${i}`,
      donors: Math.round((i * Math.random() / 2) % 20),
      // donors: 20,
      x: i,
      color: `hsl(${i % 360}, 70%, 44%)`,
    })).filter(x => x.donors > 0)),
  // {
  //   id: 'MU1',
  //   donors: 5,
  //   x: 5,
  //   color: `blue`
  // },
  // {
  //   id: 'MU2',
  //   donors: 5,
  //   x: 7,
  //   color: `blue`
  // },
  // {
  //   id: 'MU4',
  //   donors: 5,
  //   x: 95,
  //   color: `blue`
  // },
  // {
  //   id: 'MU6',
  //   donors: 5,
  //   x: 97,
  //   color: `blue`
  // }
  ],
})
