// @oncojs/react-lolliplot/dist/lib/minimap.js
'use strict';

exports.__esModule = true;

var _spatial = require('./spatial');

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var setupMinimap = function setupMinimap(props) {
  props.svg.append('g').append('rect').attrs(Object.assign({
    class: 'minimap',
    x: props.yAxisOffset,
    y: props.height - props.xAxisOffset + props.proteinHeight + 40
  }, (0, _spatial.dim)(props.xAxisLength, 50), {
    stroke: 'rgb(138, 138, 138)',
    fill: 'rgb(212, 212, 212)',
    cursor: 'text'
  })).on('mousedown', function () {
    props._update({
      draggingMinimap: true,
      zoomStart: props.d3.event.offsetX,
      offsetX: props.d3.event.offsetX
    });
  }).on('mousemove', function () {
    props._update({
      offsetX: props.d3.event.offsetX
    });
  });

  props.svg.append('g').append('clipPath').attr('id', props.uniqueSelector + '-minimap-clip').append('rect').attrs(Object.assign({
    x: props.yAxisOffset,
    y: props.height - props.xAxisOffset + props.proteinHeight + 40
  }, (0, _spatial.dim)(props.xAxisLength, 50)));

  props.svg.append('g').append('rect').attrs({
    class: 'minimap-zoom-area',
    x: props.min * props.scale + props.yAxisOffset + _spatial.halfPixel,
    y: props.height - props.xAxisOffset + props.proteinHeight + 40 + _spatial.halfPixel,
    height: 50 - 1,
    width: Math.max(1, (props.max - props.min) * props.scale - 1),
    fill: 'rgb(255, 255, 255)',
    'pointer-events': 'none'
  });

  props.svg.append('g').append('text').text('\n      Three ways to zoom in or out:\n      1) Click on a domain above.\n      2) Click and drag over the main chart above.\n      3) Click and drag over the gene map below.\n    ').attrs({
    class: 'minimap-label',
    x: props.yAxisOffset,
    y: props.height - props.xAxisOffset + props.proteinHeight + 30,
    'font-size': '11px'
  });

  props.svg.append('g').append('line').attrs({
    class: 'minimap-protein-mutation-divider',
    x1: props.yAxisOffset,
    y1: props.height - props.xAxisOffset + props.proteinHeight + 80 - _spatial.halfPixel,
    x2: props.xAxisLength + props.yAxisOffset,
    y2: props.height - props.xAxisOffset + props.proteinHeight + 80 - _spatial.halfPixel,
    stroke: _theme2.default.black
  });

  props.svg.append('g').append('text').text('aa 1').attrs({
    x: props.yAxisOffset,
    y: props.height - props.xAxisOffset + props.proteinHeight + 110,
    'font-size': '11px',
    'text-anchor': 'start'
  });

  props.svg.append('g').append('text').text('aa ' + props.domainWidth).attrs({
    x: props.width,
    y: props.height - props.xAxisOffset + props.proteinHeight + 110,
    'font-size': '11px',
    'text-anchor': 'end'
  });

  props.svg.append('g').append('text').text('This track represents the whole gene. The white area is the current zoom level.').attrs({
    x: props.width / 2,
    y: props.height - props.xAxisOffset + props.proteinHeight + 110,
    'font-size': '11px',
    'text-anchor': 'middle'
  });
};

/*----------------------------------------------------------------------------*/

exports.default = setupMinimap;
