// @oncojs/react-lolliplot/dist/lib/ticks.js
'use strict';

exports.__esModule = true;

var _lodash = require('lodash.range');

var _lodash2 = _interopRequireDefault(_lodash);

var _theme = require('./theme');

var _theme2 = _interopRequireDefault(_theme);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var setupTicks = function setupTicks(_ref) {
  var d3Root = _ref.d3Root,
      svg = _ref.svg,
      numXTicks = _ref.numXTicks,
      scaleLinearY = _ref.scaleLinearY,
      xAxisOffset = _ref.xAxisOffset,
      yAxisOffset = _ref.yAxisOffset,
      domainWidth = _ref.domainWidth,
      scale = _ref.scale,
      height = _ref.height,
      min = _ref.min,
      max = _ref.max,
      highestValue = _ref.highestValue;


  // Vertical ticks

  svg.append('g').attr('class', 'yTicks');

  d3Root.select('.yTicks').append('g').selectAll('text').data((0, _lodash2.default)(1, highestValue, highestValue / 10)).enter().append('text').text(function (i) {
    return Math.round(i);
  }).attrs({
    class: function _class(i) {
      return 'yTick-text-' + i;
    },
    x: yAxisOffset - 10,
    y: function y(i) {
      return scaleLinearY(i) + 3;
    },
    'font-size': '11px',
    'text-anchor': 'end'
  });

  d3Root.select('.yTicks').append('g').selectAll('line').data((0, _lodash2.default)(1, highestValue, highestValue / 10)).enter().append('line').attrs({
    class: function _class(i) {
      return 'yTick-line-' + i;
    },
    x1: yAxisOffset - 7,
    y1: function y1(i) {
      return scaleLinearY(i);
    },
    x2: yAxisOffset,
    y2: function y2(i) {
      return scaleLinearY(i);
    },
    stroke: _theme2.default.black
  });

  // Horizontal ticks

  svg.append('g').attr('class', 'xTicks');

  // Horizontal ticks
  var length = (max - min) / numXTicks;
  var olength = domainWidth / numXTicks;
  d3Root.select('.xTicks').append('g').selectAll('text').data((0, _lodash2.default)(numXTicks - 1).map(function (x) {
    return x + 1;
  })).enter().append('text').text(function (i) {
    return Math.round(length * i + min);
  }).attrs({
    class: function _class(i) {
      return 'xTick-text-' + i;
    },
    x: function x(i) {
      return olength * i * scale + yAxisOffset;
    },
    y: height - xAxisOffset + 20,
    'font-size': '11px',
    'text-anchor': 'middle',
    'pointer-events': 'none'
  });

  for (var i = 1; i < numXTicks; i++) {
    var _length = domainWidth / numXTicks;

    d3Root.select('.xTicks').append('line').attrs({
      class: 'xTick-line-' + i,
      x1: _length * i * scale + yAxisOffset,
      y1: height - xAxisOffset,
      x2: _length * i * scale + yAxisOffset,
      y2: height - xAxisOffset + 10,
      stroke: _theme2.default.black,
      'pointer-events': 'none'
    });
  }
};

/*----------------------------------------------------------------------------*/

exports.default = setupTicks;
